### PR TITLE
CRM-20906 Validate the extension key is of general standard

### PR DESCRIPTION
**Pull request by seamuslee001**
Overview
----------------------------------------
Previously there was no checking if the key passed in through the URL was a standard extension key. This adds some validation

ping @jackrabbithanna @totten @eileenmcnaughton @seanmadsen

---

 * [CRM-20906](https://issues.civicrm.org/jira/browse/CRM-20906)

### CODE REVIEW

**Review by seancolsen — CHANGES_REQUESTED** [review]
I have reviewed locally and requested a change via https://github.com/seamuslee001/civicrm-core/pull/5 

@seamuslee001 sorry I gave you a regex via Mattermost without fully testing it. Minor oversight here. 

Before my added commit, I was still able to reproduce the problem described in [CRM-20906](https://issues.civicrm.org/jira/browse/CRM-20906). With the addition of my latest commit I was no longer able to repro.

**Comment by seanmadsen** [comment]
Thanks @seamuslee001 

This looks good to me now on all fronts. Ready to merge :white_check_mark:

**Comment by seanmadsen** [comment]
Great that you've added a unit test. 

But now that you've done that I'm going to surface a nit-pick thought that I originally suppressed in my review... 

I think it would be nice to put the logic which validates an extension key into `CRM_Utils_Rule` so that it can be used elsewhere as needed. Then in the form layer, instead of `self::checkExtesnionKeyIsValid($this->_key)` we could do `CRM_Utils_Type::validate($this->_key, 'ExtensionKey')`. And the unit tests go along with `CRM_Utils_Rule`. What do you think of this approach?

**Comment by seamuslee001** [comment]
O_o that could work i guess

**Comment by seanmadsen** [comment]
I general, I'd love to see us add lots of other types to `CRM_Utils_Rule`, for example, `CommaSeparatedIntegers`, so that we can move away from validating inputs simply as "string". Moving this logic there in this PR would be a step in that direction.

**Comment by seanmadsen** [comment]
This looks great. Even better than before! Tested locally. Confirmed to fix issue. Code looks good. Ready to merge :white_check_mark:

**Review by seancolsen — APPROVED** [review]

**Comment by seamuslee001** [comment]
@colemanw i think this is right for merging now and has been tested by Sean

File: CRM/Utils/Rule.php [hunk 1]
```diff
@@ -911,4 +911,15 @@ public static function validDateRange($fields, $fieldName, &$errors, $title) {
     }
   }
 
+  /**
+   * @param string $key Extension Key to check
+   * @return bool
+   */
+  public static function checkExtesnionKeyIsValid($key = NULL) {
```
Comment (eileenmcnaughton) [inline CRM/Utils/Rule.php]: Too many beers when writing thish function shignature

**Pull request merged**